src/pip/_internal/exceptions.py

```diff
@@ -22,75 +24,140 @@ def _is_kebab_case(s: str) -> bool:
+    """An error, that presents diagnostic information to the user.
 
     This contains a bunch of logic, to enable pretty presentation of our error
     messages. Each error gets a unique reference. Each error can also include
     additional context, a hint and/or a note -- which are presented with the
     main error message in a consistent style.
+
+    This is adapted from the error output styling in `sphinx-theme-builder`.
     """
 
     reference: str
 
     def __init__(
         self,
         *,
-        message: str,
-        context: Optional[str],
-        hint_stmt: Optional[str],
-        attention_stmt: Optional[str] = None,
-        reference: Optional[str] = None,
         kind: 'Literal["error", "warning"]' = "error",
+        reference: Optional[str] = None,
+        message: Union[str, Text],
+        context: Optional[Union[str, Text]],
+        hint_stmt: Optional[Union[str, Text]],
+        note_stmt: Optional[Union[str, Text]] = None,
+        link: Optional[str] = None,
     ) -> None:
-
         # Ensure a proper reference is provided.
         if reference is None:
             assert hasattr(self, "reference"), "error reference not provided!"
             reference = self.reference
         assert _is_kebab_case(reference), "error reference must be kebab-case!"
 
-        super().__init__(f"{reference}: {message}")
-
         self.kind = kind
+        self.reference = reference
+
         self.message = message
         self.context = context
 
-        self.reference = reference
-        self.attention_stmt = attention_stmt
+        self.note_stmt = note_stmt
         self.hint_stmt = hint_stmt
 
-    def __str__(self) -> str:
-        return "".join(self._string_parts())
-
-    def _string_parts(self) -> Iterator[str]:
-        # Present the main message, with relevant context indented.
-        yield f"{self.message}\n"
-        if self.context is not None:
-            yield f"\n{self.context}\n"
+        self.link = link
 
-        # Space out the note/hint messages.
-        if self.attention_stmt is not None or self.hint_stmt is not None:
-            yield "\n"
+        super().__init__(f"<{self.__class__.__name__}: {self.reference}>")
 
-        if self.attention_stmt is not None:
-            yield _prefix_with_indent("Note: ", self.attention_stmt)
+    def __repr__(self) -> str:
+        return (
+            f"<{self.__class__.__name__}("
+            f"reference={self.reference!r}, "
+            f"message={self.message!r}, "
+            f"context={self.context!r}, "
+            f"note_stmt={self.note_stmt!r}, "
+            f"hint_stmt={self.hint_stmt!r}"
+            ")>"
+        )
 
+    def __rich_console__(
+        self,
+        console: Console,
+        options: ConsoleOptions,
+    ) -> RenderResult:
+        colour = "red" if self.kind == "error" else "yellow"
+
+        yield f"[{colour} bold]{self.kind}[/]: [bold]{self.reference}[/]"
```

Comment:
Be careful with f strings and console markup! If `self.kind` or `self.reference` contain square brackets, they may be interpreted as tags.

You can either use `escape` from `rich.markup`, or use Text class directly.

With escape:

```python
yield f"[{colour} bold]{escape(self.kind)}[/]: [bold]{escape(self.reference)}[/]"
```

With Text class:

```python
yield Text.assemble(
    (self.kind, Style.parse(colour) + Style(bold=True)),
    ": ",
    (self.reference, "bold")
)
```

Reply:
Both of these are required to not contain [ (one is a literal "error", the other is required to be kebab case).